### ISSUE TITLE
Gesture Animation is not 120hz and feels laggy on ProMotion iPhone 13

### ISSUE THREAD

**Issue by eric8810** [flutter/flutter#101653]
### Details

with https://github.com/flutter/flutter/issues/90675
the interface refresh rate should be fixed to 120hz on Iphone 13 promotion screen.
the problem is , if you drag list or screen with fingers, the ui is definitly running at 60hz.

### 1.  how to reproduce
see difference between drag list view slowly **VS** drag & release list view fast 

### 2. demo repo
https://github.com/eric8810/gesture-animation-demo

list drag video:
https://drive.google.com/file/d/1LGwoQUQ2PmlIfS41W5kvE-bYT6SNoWyX/view?usp=sharing

navigation drag video:
https://drive.google.com/file/d/1rp-exZ0RGs34A9q26rGAE2W9yzoLZ3Q9/view?usp=sharing

**Target Platform: iOS
**Target OS version/browser: 15.4.1
**Devices: iPhone 13 Pro Max

## Traceing file

**Timeline tracing file:**

[dart-timeline-2022-4-14.json.zip](https://github.com/flutter/flutter/files/8485980/dart-timeline-2022-4-14.json.zip)

## Logs

<details>

```
[✓] Flutter (Channel master, 2.13.0-0.0.pre.438, on macOS 12.3.1 21E258 darwin-arm, locale zh-Hans-CN)
    • Flutter version 2.13.0-0.0.pre.438 at /Users/eric8810/code/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision e3f8b50fd5 (2 days ago), 2022-04-07 20:51:07 -0400
    • Engine revision 816e3d3012
    • Dart version 2.17.0 (build 2.17.0-266.0.dev)
    • DevTools version 2.12.1

[✓] Android toolchain - develop for Android devices (Android SDK version 32.1.0-rc1)
    • Android SDK at /Users/eric8810/Library/Android/sdk
    • Platform android-32, build-tools 32.1.0-rc1
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7772763)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7772763)

[✓] VS Code (version 1.67.0-insider)
    • VS Code at /Applications/Visual Studio Code - Insiders.app/Contents
    • Flutter extension version 3.39.20220405

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 12.3.1 21E258 darwin-arm
    • Chrome (web)    • chrome • web-javascript • Google Chrome 100.0.4896.75

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```
</detail>

**Comment by eric8810** [comment]
#90675

**Comment by maheshmnj** [comment]
Hi @eric8810, Thanks for filing the issue. Labeling this issue for further investigation from the team.

cc: @cyanglaz

**Comment by zanderso** [comment]
cc @iskakaushik

**Comment by acoutts** [comment]
> #90675

This is a bit different - while we indeed have 120hz for animations and flings during scrolling now, it appears it's true while you are in the act of dragging the scroll, it's only 60hz.

**Comment by eric8810** [comment]
> Hi @eric8810, Thanks for filing the issue. Since you have the reproducible device, Can you please attach the timeline trace JSON file? You can find the instructions for it in this issue #74833 Labeling this issue for further investigation from the team.
> 
> cc: @cyanglaz

timeline json has been uploaded

**Comment by roybeitner** [comment]
Having this issue as well, scrolling overall on iOS is laggy and doesn't feel smooth like native apps.
With 120hz enabled, small scrolls are smoother but you have more noticeable lag overall.

**Comment by jmagman** [comment]
@cyanglaz @iskakaushik any chance either of you had a chance to look at this?

**Comment by cyanglaz** [comment]
> @cyanglaz @iskakaushik any chance either of you had a chance to look at this?

I haven't, but it is in my queue.

The next step (if someone would like to try before I do) is to check what the frame rate budget the CADisplayLink provides during those frames:
To quickly check this, you can build a local engine and check the value of `current_refresh_rate_` during those laggy frames here: https://github.com/flutter/engine/blob/main/shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.mm#L116

If this value becomes lower, it is actually the platform(iOS) decided to lower the frame rate budget. From there, we can figure out when and why it happens and what we can do to either 1. tell iOS system to ramp up the frame rate in those scenarios, or 2. to make flutter doesn't seem as laggy while a lower frame rate is provided.

**Comment by cyanglaz** [comment]
Update: I tried it on my iPhone 13 pro iOS 15.1, the refresh rate that reported (using the above method) is always 120Hz and it also looks smooth. I'm going to try on iOS 15.4.1.

**Comment by rotorgames** [comment]
I have the same issue. iPhone 13 pro 15.4.1

I found that:
When you scroll a list by your finger, moving looks like 60hz (but the performance profiler shows 120fps).
As soon as you scroll and stop touching the screen by your finger, it looks smoother like 120hz.
But as soon as you touch the screen again, all moves become 60hz again (but the performance profiler shows 120fps).

Looks like all animations works with 120hz but all gestures update with 60hz. 

My opinion there is a problem with gestures.

**Comment by eric8810** [comment]
> Update: I tried it on my iPhone 13 pro iOS 15.1, the refresh rate that reported (using the above method) is always 120Hz and it also looks smooth. I'm going to try on iOS 15.4.1.

yes, the reported refresh rate is 120hz
but animation with guestures are running at 60hz, even the whole screen is at 120hz

**Comment by cyanglaz** [comment]
@rotorgames @eric8810 May I ask I do you tell the screen is running at 60hz instead of 120hz while dragging the list view. It seems like my eyes are not good enough to tell 60hz from 120hz and it looks smooth to me :p 
Is there a way to actually see some numbers?

**Comment by rotorgames** [comment]
@cyanglaz That's sounds strange but I can't replicate this issue with the debug configuration. But as soon as I change the configuration to release mode I see this issue again.

So, try to build the app with release configuration.

**Comment by eric8810** [comment]
> @rotorgames @eric8810 May I ask I do you tell the screen is running at 60hz instead of 120hz while dragging the list view. It seems like my eyes are not good enough to tell 60hz from 120hz and it looks smooth to me :p Is there a way to actually see some numbers?

I didnt get any useful number, or do you have other suggestions?
I have upload an slow motion video for you

**Comment by rotorgames** [comment]
@cyanglaz Any updates?

**Comment by cyanglaz** [comment]
Since the refresh rate reported by the platform is 120hz, the only thing I can think of is that each frame takes too long to render, it seems unlikely because @rotorgames mentioned that it wasn't reproducible with the debug mode. 

@iskakaushik Do you want to give it a try and see if you can reproduce it?

**Comment by rotorgames** [comment]
@cyanglaz I don't think that each frame takes too long to render, because we get truthful 120hz for scroll animation when we don't touch the screen but as soon as we touch the screen and move a finger to scroll the list (or other widgets that react on gestures) we get 60hz (or even less, i don't know). It looks like that gestures notify UI not 120hz per second.

**Comment by rotorgames** [comment]
@cyanglaz I did a deep digging and found some strange things. 

### First
Sometimes, it starts to work fine, I don't know why, but if I keep my iPhone 13 Pro connected to my mac for a long time, the scroll starts to be 120hz but as soon as I detach my device from the mac, it starts to work with a slow rate again. Sometime I can replicate it, sometime I can't.

### Second
I profiled my simple app in the Xcode Instruments and I found that FPS for "finger scroll" and "inertial scroll" are really different. "finger scroll" is not 60hz, it's around 70-80hz but it's not 120hz. 

<img width="1290" alt="profiler" src="https://user-images.githubusercontent.com/1790504/169432932-a71d2971-00ff-46e4-973d-de5cc4365418.png">

Also, I found that during "finger scrolling" UI thread is blocking for 15.21ms but for "inertial scrolling" the thread is blocking for 5.18ms. It's 3 times less.

<img width="1289" alt="profiler2" src="https://user-images.githubusercontent.com/1790504/169433630-71eb9d35-4ecb-4755-a365-33581e489a90.png">

So, as result, I guess the Flutter render engine works fast enough, but `GestureDetector (or something that is responsible for notifying system that widgets must be re-rendered)` doesn't notify the flutter engine that widgets must be re-rendered as fast as the system allows.

The project I used to replicate the issue: https://github.com/rotorgames/flutter_scroll_fps_test

* iPhone 13 Pro (iOS 15.4.1)
* MacOS Montery 12.3.1
* Xcode 13.3.1
* Flutter 3.0.0 (the same issue with 2.x as well) 
* Build with a release mode

**Comment by rotorgames** [comment]
More info:

I found that `_handlePointerDataPacket` method in `flutter/packages/flutter/lib/src/gestures/binding.dart` is not invoking with 120hz as well. I guess it's the first method that is running on the Flutter's side. All methods prior `_handlePointerDataPacket` are in the Dart or iOS side. So, looks like iOS doesn't want to notify a flutter app with 120hz during dragging.

IMHO

**Comment by WithToken** [comment]
This problem is more obvious on other phones. 
iPhone 12

**Comment by rotorgames** [comment]
So, I guess I found the source of the problem. 

### Why do we have lags? 
The reason is that we pause `CADisplayLink`. Screens Frame Rate and Touch Frame Rate (touchesMoved) in iOS are connected with each other. As soon as we pause `CADisplayLink`, iOS doesn't need to call `touchesMoved` 120 times per second anymore, so, iOS run `CADisplayLink` with 60hz but, as soon as `touchesMoved` is called we unpause `CADisplayLink` and iOS see that there is a `CADisplayLink` with 120 frame rate, then iOS try to run this `CADisplayLink` with 120 frame rate, but in `onDisplayLink` we pause the `CADisplayLink` again and this behavior is looping.

### Why don't we see this issue with animations?
Because animations are not depended on a touch screen. They do their work each frame without gaps, so iOS doesn't checkout to 60hz because of it (IMHO)

### Is it a Flutter bug?
I guess this is unexpected behavior on iOS side. Apple thought that we would start `CADisplayLink` when we need and pause or invalidate when we done all work. But Flutter pauses `CADisplayLink` each frame. I guess `CADisplayLink` is not ready for that.

### How to fix?
We can unpause `CADisplayLink` when `touchesBegan` is calling and pause when `touchesEnded` or `touchesCancelled` are calling. We don't have to pause `CADisplayLink` during a user touches the screen. Then we can get perfect and smooth work on iOS devices.

Also, I could get 120hz only when I added `CADisplayLink` into `mainRunLoop` and disabled pausing. 

As a measurement result I got a true 120hz on a device and perfect work without any freezes and lags.

If you disagree with my conclusion, let's discuss it. Anyway, we have to fix this issue asap because apps on a device with $2000 price must work perfect, with no lags end etc.

**Comment by eric8810** [comment]
any progress? should we moving on?

**Comment by ollydixon** [comment]
Hey guys; just chipping in here. I think this is something we need to prioritize; right now our Flutter apps have a very low quality feeling to them on iOS and MacOS devices.

**Comment by sundayla** [comment]
Hi, my case is with the textfield soft keyboard animation.

My app included:
Webview (video embed)
Listview (as chatroom message)
textfield 

It is tested on iPhone 13 Pro Max, release mode:
- When the keyboard showup, the listview and textfield will pushup normally but not smooth.  
Then I play the video inside my webview
- The keyboard animation change to very smooth (I think it is casued by the playing vidoe change something about refresh reate?)

Also tested on low power mode or record it as video, when the screen locked to 60hz, the animation is not that bad.

**Comment by cyanglaz** [comment]
I have run the demo app with instrument in profile mode, the refresh rate is pretty constant at 8ms, whether I have my fingers touching the screen while scrolling or I do a quick scroll with finger touch down and up. Here is a screenshot of the instrument result:
<img width="1492" alt="Screen Shot 2022-07-14 at 10 55 16 AM" src="https://user-images.githubusercontent.com/3756895/179050917-c55a6e5e-5031-4b9b-ad55-8771e54f577f.png">

The little bump on the screenshot are 9ms, which is expected and acceptable as iOS sometimes lower the refresh rate for different reasons. 

The displayLink.duration also almost always shows 8ms as the duration. 

In summary, I haven't observe any significant refresh rate increase during the scroll gesture. 

I tested on iPhone 13 pro and iOS 15.4.1

**Comment by cyanglaz** [comment]
So I tried to scroll the list view in the 2nd list page with automated animation. It also creates a 16ms frame at the start of the each scrolling. So I believe it is not related to the touch events. 
I also tried automatic scrolling on the list view of the main page. All frames are 8ms.

Knowing the difference between the main page and 2nd page would help us find the root cause of this issue.

**Comment by ollydixon** [comment]
@cyanglaz hey; they know the issue; some widgets are locked to 60 frames for technical  reasons (probably isn’t using an delta double in the render loop)

Just a case of them unlocking it and doing a lot of widget testing 

:-)

**Comment by rotorgames** [comment]
@cyanglaz Iphone 13 Pro iOS 15.5. I still have this issue. All animations work fine, I see the issue only if I touch the screen. I don't see janks only if I don't pause `CADisplayLink` in [vsync_waiter_ios.mm:119](https://github.com/flutter/engine/blob/main/shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.mm#L119)

Could you try another device with ProMotion screen? Maybe you can't replicate this issue because of your device or device's settings.

**Comment by ollydixon** [comment]
@rotorgames lets wait for them to fix this; it’s an obvious issue at the moment. For example ListView is locked to 60fps.

**Comment by JagandeepBrar** [comment]
For another data point, my application was running at 120hz without issue on my iPhone 13 Pro Max on iOS 15.5.

I recently upgraded my device to iOS 16 and am now experiencing the same issues where animations are clearly running at 120hz but ListViews and PageViews are now noticeable running at 60hz.

Tested the exact same build (deployed to TestFlight) on my iPad Pro and everything within the application is smoothly running at 120hz.

**Comment by cyanglaz** [comment]
Here is my conclusion after testing and researching.
1. Flutter engine pauses vsync waiter when there is no new frame to draw. It seems like the first frame after a long pause usually takes longer to render, which results a 60hz frame. This is the bug for both navigation animation and scrolling animation mentioned in this issue. The issue is not related to gestures. I'm not sure if the framework has some mechanism to lock the refresh rate at the start of an animation. cc @chunhtai 
2. Although it is sometimes OK to have one frame below 120hz. Some app needs the 120hz smooth feeling. We can workaround by force asking the framework to render every frame. For scroll:
```dart
_ticker = createTicker((elapsed) {
  setState(() {});
});
NotificationListener<ScrollNotification>(
  onNotification: (notification) {
    if (notification is ScrollStartNotification) {
      _ticker.start();
    }
    if (notification is ScrollEndNotification) {
      _ticker.stop();
    }
    return false;
  },
  child: ListView.builder(
    itemCount: itemList.length,
    itemBuilder: (context, index) {
      return ListTile(
        title: Text(itemList[index]),
        tileColor: getColor(index),
      );
    },
  ),
),
```
A similar approach can probably be applied to the navigation animation. I'm not familiar with how to listen to navigation animation.

**Comment by rotorgames** [comment]
@cyanglaz the solution with a ticker works but not perfect. I can get 120fps but sometimes when I scroll the list the first time after a could boot I still get 60 fps. When I scroll the list the second time it's fine.

**Comment by cyanglaz** [comment]
> @cyanglaz the solution with a ticker works but not perfect. I can get 120fps but sometimes when I scroll the list the first time after a could boot I still get 60 fps. When I scroll the list the second time it's fine.

What is a "could boot", could you explain how I can reproduce this?

**Comment by chunhtai** [comment]
I am not familiar with how frame rate is controlled, or can it even be controlled, from the framework side.
maybe @goderbauer knows?

**Comment by rotorgames** [comment]
@cyanglaz I meant cold boot. When the app is removed from the memory and open again.

**Comment by rockingdice** [comment]
Bump this. We are receiving more reports from our customers that the issue is reproducible on iOS 15.5 devices of the latest models. iPhone 7 (and old models) is not reproducible.
But my situation is a bit different, we are still using flutter v2.2 stable.
This should be fixed asap. The customers are quite confusing why their newly bought devices are laggy.

**Comment by cyanglaz** [comment]
> Bump this. We are receiving more reports from our customers that the issue is reproducible on iOS 15.5 devices of the latest models. iPhone 7 (and old models) is not reproducible. But my situation is a bit different, we are still using flutter v2.2 stable. This should be fixed asap. The customers are quite confusing why their newly bought devices are laggy.

Enabling 120hz is introduced in https://github.com/flutter/flutter/commit/0b97874895eb3e14c1cf0e65bd7fca3a17c62b02, which is not available in flutter v2.2. If you need this feature, please upgrade flutter to at least 3.0.0

**Comment by rockingdice** [comment]
> 

I don't really need to support 120hz but should it be smoothing on 60hz as well? My customers complain that the list dragging is laggy, and I cannot reproduce it on a 60hz device. All issued devices are 120hz.
Maybe I'm off-topic on this thread, is there an existing issue for my situation?

**Comment by zanderso** [comment]
@rockingdice Please file a new issue that includes a timeline trace and a video from a profile build that shows what is laggy.

**Comment by iosephmagno** [comment]
Hello! Any progress on this? That is severe issue and we will be delaying our app launch due to it.

**Comment by zanderso** [comment]
@cyanglaz Regarding:

> It seems like the first frame after a long pause usually takes longer to render, which results a 60hz frame

Do you have timelines that you can share that demonstrate this? This sounds like something that we should dig into.

**Comment by cyanglaz** [comment]
@zanderso Here are the dart-timeline and dart_devtool exported jsons. 

[Archive.zip](https://github.com/flutter/flutter/files/9245935/Archive.zip)

**Comment by zanderso** [comment]
Looks like the raster thread is getting descheduled, but I don't know why:
<img width="606" alt="Screen Shot 2022-08-03 at 8 56 18 AM" src="https://user-images.githubusercontent.com/6343103/182654746-5a6e0c3d-3d01-4c69-8227-dcf102947cd0.png">

Doesn't look like there's anything else going on there. Some, but not all, of the long frames have a parallel scavenge happening at the same time, but that doesn't look like the reason the raster thread is getting kicked off the CPU. Maybe others have more of a clue than me. @iskakaushik or @dnfield maybe.

**Comment by luckysmg** [comment]
Hi @cyanglaz, I raised a PR for fixing this issue.
Would you mind take a look when you have a chance?😄😄

**Comment by rotorgames** [comment]
@luckysmg Tested up on my iPhone 13 Pro iOS 15.6

Works like a charm.

**Comment by b3nni97** [comment]
@luckysmg Does your PR also help for the lags of the devices with 60hz?

**Comment by luckysmg** [comment]
Hi @b3nni97 
Drag behavior on 60HZ devices should be have no lags and it is smooth originally. 
That PR is for adjustment of promotion devices.

**Comment by b3nni97** [comment]
@luckysmg Ok then this seems like a different issue #103847

**Comment by luckysmg** [comment]
Fixed by https://github.com/flutter/engine/pull/35592

**Comment by iosephmagno** [comment]
@luckysmg Great job! 👏 Can you please have a look at #103847 and share your thoughts on that?

**Comment by luckysmg** [comment]
@iosephmagno 
Seems it's a different issue. I will take a look later when I have a chance....😄
